packages/stores/src/account/osmosis/index.ts

```diff
@@ -1605,7 +1606,223 @@ export class OsmosisAccountImpl {
+      const querySharePool = this.queries.queryPools.getPool(
+        shares.denom.split("/")[2]
+      );
+
+      // unexpected pool type
+      if (!querySharePool || !querySharePool.sharePool)
+        throw new Error(`Pool ${shares.denom.split("/")[2]} not found`);
+
+      // save for later
+      involvedQueryPools.push(querySharePool);
+
+      // update pool data
+      await querySharePool.waitFreshResponse();
+
+      /// 2
+
+      const estimated = OsmosisMath.estimateExitSwap(
+        {
+          totalShare: querySharePool.sharePool.totalShare,
+          poolAssets: querySharePool.poolAssets.map(({ amount }) => ({
+            denom: amount.toCoin().denom,
+            amount: new Int(amount.toCoin().amount),
+          })),
+          exitFee: querySharePool.exitFee.toDec(),
+        },
+        this.makeCoinPretty,
+        new CoinPretty(querySharePool.shareCurrency, shares.amount)
+          .toDec()
+          .toString(),
+        querySharePool.shareCurrency.coinDecimals
+      );
+
+      const underlyingShareCoins: CoinPrimitive[] = estimated.tokenOuts.map(
+        (tokenOut) => ({
+          denom: tokenOut.currency.coinMinimalDenom,
+          amount: tokenOut
+            .toDec()
+            .mul(new Dec(1).sub(maxSlippageDec))
+            .mul(
+              DecUtils.getTenExponentNInPrecisionRange(
+                tokenOut.currency.coinDecimals
+              )
+            )
+            .truncate()
+            .toString(),
+        })
+      );
+
+      /// 3, 4
+
+      const swapPromises: Promise<Int>[] = underlyingShareCoins.map((coin) => {
+        if (coin.denom === stakeCurrency.coinMinimalDenom)
+          return Promise.resolve(new Int(coin.amount));
+        else {
+          // swap this non-stake currency for the stake currency out
+          const token = { ...coin, amount: new Int(coin.amount) };
+
+          return new Promise((resolve, reject) =>
+            querySharePool.pool
+              .getTokenOutByTokenIn(token, stakeCurrency.coinMinimalDenom)
+              .then((quote) => resolve(quote.amount))
+              .catch(reject)
+          );
+        }
+      });
+
+      return await Promise.all(swapPromises).then((swaps) =>
+        swaps.reduce((acc, swap) => acc.add(swap), new Int(0))
+      );
+    };
+
+    const msgPromises: Promise<EncodeObject>[] = convertibleAssets
+      .map((asset) => {
+        if ("lockId" in asset) {
+          // share pool shares in lock
+
+          const userPeriodLock = queryAccountLocked.getPeriodLockById(
+            asset.lockId
+          );
+          const poolAssetCoin = userPeriodLock?.coins.find((coin) =>
+            coin.denom.includes("/pool/")
+          );
+
+          // this lock contains some unexpected asset
+          if (!poolAssetCoin)
+            throw new Error(
+              `Lock ID ${asset.lockId} contains some unexpected asset`
+            );
+
+          if (poolAssetCoin.denom.includes("gamm/pool/")) {
```

Comment:
We should probably declare 'magic strings' as constants.

Reply:
Where should the constants live?